lib/Service/CategoriesService.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * Calendar App
+ *
+ * @copyright 2023 Claus-Justus Heine <himself@claus-justus-heine.de>
+ *
+ * @author Claus-Justus Heine <himself@claus-justus-heine.de>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Calendar\Service;
+
+use OCP\Calendar\ICalendar;
+use OCP\Calendar\ICalendarQuery;
+use OCP\Calendar\IManager as ICalendarManager;
+use OCP\IL10N;
+use OCP\SystemTag\ISystemTag;
+use OCP\SystemTag\ISystemTagManager;
+use Psr\Log\LoggerInterface;
+
+class CategoriesService {
+	/** @var null|string */
+	private $userId;
+
+	/** @var ICalendarManager */
+	private $calendarManager;
+
+	/** @var ISystemTagManager */
+	private $systemTagManager;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	/** @var IL10N */
+	private $l;
+
+	private const CALENDAR_OBJECT_PROPERTIES_TABLE = 'calendarobjects_props';
+
+	public function __construct(?string $userId,
+		ICalendarManager $calendarManager,
+		ISystemTagManager $systemTagManager,
+		LoggerInterface $logger,
+		IL10N $l10n) {
+		$this->userId = $userId;
+		$this->calendarManager = $calendarManager;
+		$this->systemTagManager = $systemTagManager;
+		$this->logger = $logger;
+		$this->l = $l10n;
+	}
+
+	/**
+	 * This is a simplistic brute-force extraction of all already used
+	 * categories from all events accessible to the currently logged in user.
+	 *
+	 * @return array
+	 */
+	private function getUsedCategories(): array {
+		if (empty($this->userId)) {
+			return [];
+		}
+
+		$categories = [];
+		$principalUri = 'principals/users/' . $this->userId;
+		$query = $this->calendarManager->newQuery($principalUri);
+		$query->addSearchProperty(ICalendarQuery::SEARCH_PROPERTY_CATEGORIES);
+		$calendarObjects = $this->calendarManager->searchForPrincipal($query);
+		foreach ($calendarObjects as $objectInfo) {
+			foreach ($objectInfo['objects'] as $calendarObject) {
+				if (isset($calendarObject['CATEGORIES'])) {
+					$eventCategories = explode(',', $calendarObject['CATEGORIES'][0][0]);
+					$categories = array_merge(
+						$categories,
+						array_fill_keys($eventCategories, true)
+					);
+				}
+			}
+		}
+		$categories = array_filter(array_map(fn ($label) => trim($label), array_keys($categories)));
```

Comment:
could you add a comment explaining what the filter does here?

Reply:
I experience some "broken" categories in some instances which are surrounded by spaces or even completely "empty". I wanted to avoid to inject those into the category multi-select in the UI. I have added a comment.